### ISSUE TITLE
App's Functional Tests in the 3.0 Directory structure

### ISSUE THREAD

**Issue by ErichHartmann** [symfony/symfony-standard#771]
As a minor observation and for your consideration, I have noticed my need to move functional tests (e.g. WebTestCase) in to an /app/tests directory and treat them as exclusive to each app. As I began to reuse bundles more in new projects, I realized my functional tests located inside the bundles' directories were affected by new app-level routing. In particular, the the use of prefixes in routing.yml configurations created a brittle connection to the web test cases inside the bundles.

For example, here is the standard web test case provided in the docs and produced by the bundle generator:
```php
// src/Acme/DemoBundle/Tests/Controller/DemoControllerTest.php
namespace Acme\DemoBundle\Tests\Controller;

use Symfony\Bundle\FrameworkBundle\Test\WebTestCase;

class DemoControllerTest extends WebTestCase
{
    public function testIndex()
    {
        $client = static::createClient();

        $crawler = $client->request('GET', '/demo/hello/Fabien');

        $this->assertGreaterThan(0, $crawler->filter('html:contains("Hello Fabien")')->count());
    }
}
```
However, changes in the app's routing configuration such as the below example cause a test failure.
```yml
# app/config/routing.yml
_acme_demo:
    resource: "@AcmeDemoBundle/Controller/"
    type:     annotation
    prefix:   /Acme/
```
I also note this advisory in the docs:
```
Hardcoding the request URLs is a best practice for functional tests. If the test
generates URLs using the Symfony router, it won't detect any change made to
the application URLs which may impact the end users.
```

Base on my experiences, if I didn't treat the web test cases as exclusive to each app, I would end up modifying each bundle for each app.

**Comment by stof** [comment]
Well, I see 2 things here:

- functional tests for app-specific bundles (the AppBundle when following best practices) are tied to the app (they are actually functional tests for the app itself). I think we could even put the tests in a ``tests`` folder at the root (near ``src``)
- functional tests for shared bundles should be standalone (embedding a fixture app in their testsuite, as done in FrameworkBundle and SecurityBundle for instance)

**Comment by ErichHartmann** [comment]
@stof 

Ok; I'm not suggesting a debate about an app directory versus just a src directory again.

When we gave up creating views inside bundles and moved them to the app/resource directories it made a lot more sense me in terms of the portability of bundles. I made the same intellectual leap with functional tests.

Keeping the app's web test cases outside the bundles allowed me to finally maintain distinct source control with respect to bundles.
